src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs

```diff
@@ -106,26 +125,45 @@ public class CorsMiddleware
                 var corsPolicy = _policy ?? await _corsPolicyProvider?.GetPolicyAsync(context, _corsPolicyName);
                 if (corsPolicy != null)
                 {
-                    var corsResult = _corsService.EvaluatePolicy(context, corsPolicy);
-                    _corsService.ApplyResult(corsResult, context.Response);
-
                     var accessControlRequestMethod =
                         context.Request.Headers[CorsConstants.AccessControlRequestMethod];
                     if (string.Equals(
                             context.Request.Method,
                             CorsConstants.PreflightHttpMethod,
                             StringComparison.OrdinalIgnoreCase) &&
-                            !StringValues.IsNullOrEmpty(accessControlRequestMethod))
+                        !StringValues.IsNullOrEmpty(accessControlRequestMethod))
                     {
+                        ApplyCorsHeaders(context, corsPolicy);
+
                         // Since there is a policy which was identified,
                         // always respond to preflight requests.
                         context.Response.StatusCode = StatusCodes.Status204NoContent;
                         return;
                     }
+                    else
+                    {
+                        context.Response.OnStarting(c => {
+                            try
+                            {
+                                ApplyCorsHeaders((HttpContext)c, corsPolicy);
```

Comment:
Pass both parameters in via state like this: https://github.com/aspnet/KestrelHttpServer/pull/2670/files#diff-4ab77515238e5e964ae9a3a4615205b1R713
Then this callback can become a cached instance.